File: src/main/java/com/github/liebharc/queryenrichment/Selector.java
package com.github.liebharc.queryenrichment;

import java.util.Optional;

/**
 * A selector tells a {@link Query} which properties or columns should be selected.
 * @param <TAttribute> Attrigbute type
 */
public class Selector<TAttribute> implements ExecutableStep<TAttribute, Object> {

    private static final long serialVersionUID = -7538684432852388470L;
    /** The attribute which is set with the given selector */
    private final Attribute<TAttribute> attribute;
    /** The column or property name which must be queried for */
    private final String column;
    /** A dependency, normally selectors have no dependencies */
    private final Dependency dependency;

    public Selector(Attribute<TAttribute> attribute, String columnOrNull) {
       this(attribute, columnOrNull, Dependencies.noDependencies());
    }

    public Selector(Attribute<TAttribute> attribute, String columnOrNull, Dependency dependency) {
        this.attribute = attribute;
        this.column = columnOrNull;
        this.dependency = dependency;
    }

    @Override
    public Optional<String> getColumn() {
        return Optional.ofNullable(column);
    }

    @Override
    public Attribute<TAttribute> getAttribute() {
        return attribute;
    }

    @Override
    public final void enrich(IntermediateResult result, Object parameter) {
        this.enrich(result);
    }

    public void enrich(IntermediateResult result) {
        result.addFromQuery(this);
    }

    @Override
    public Dependency getDependencies() {
        return dependency;
    }

    @Override
    public final boolean isConstant() {
        return false;
    }

    @Override
    public String toString() {
        return "Step{" +
                "attribute=" + attribute +
                '}';
    }
}

File: src/main/java/com/github/liebharc/queryenrichment/Selector.kt
package com.github.liebharc.queryenrichment

import java.util.Optional

/**
 * A selector tells a [Query] which properties or columns should be selected.
 * @param <TAttribute> Attrigbute type
</TAttribute> */
open class Selector<TAttribute> @JvmOverloads constructor(
        /** The attribute which is set with the given selector  */
        override val attribute: Attribute<TAttribute>,
        /** The column or property name which must be queried for  */
        override val column: String?,
        /** A dependency, normally selectors have no dependencies  */
        override val dependencies: Dependency = Dependencies.noDependencies()) : ExecutableStep<TAttribute, Any?> {

    override val canBeConstant: Boolean
        get() = false

    override fun enrich(result: IntermediateResult, parameter: Any?) {
        this.enrich(result)
    }

    fun enrich(result: IntermediateResult) {
        result.addFromQuery(this)
    }

    override fun toString(): String {
        return "Step{" +
                "attribute=" + attribute +
                '}'.toString()
    }

    companion object {

        private val serialVersionUID = -7538684432852388470L
    }
}
